src/gpu/intel/jit/eltwise_injector.cpp

```diff
@@ -306,6 +307,56 @@ void eltwise_injector_f32_t<ngen_generator_t>::round_compute_fwd(
     h->rnde(simd, r, r);
 }
 
+template <typename ngen_generator_t>
+void eltwise_injector_f32_t<ngen_generator_t>::mx_scale_compute_fwd(int simd,
+        const ngen::GRF &r, int off, const ngen::Subregister &seed, int nreg,
+        const ngen::DataType dst_dt, int phase) {
+    assert(simd == 32);
+    assert(dst_dt == ngen::DataType::bf8);
+    ngen::GRF r_alt(off);
+    auto max = scratch_[0].f();
+    ngen::GRF tmp;
+    if (hw() >= gpu_xe_hpc) {
+        tmp = scratch_[1].f();
+    } else {
+        tmp = scratch_[2].f();
+    }
+    auto fmax = (dst_dt == ngen::DataType::bf8) ? Immediate::f(57344)
+                                                : Immediate::f(6);
+    auto scale_dst = seed.ub((phase / nreg) * 4);
+    h->template mov<float>(16, max, Immediate::f(0));
+    h->template mov<float>(16, tmp, Immediate::f(0));
+
+    // Get Max value from group.
+    h->sel(16 | ge, max.f(0)(1), abs(r.f(0)(1)), abs(r_alt.f(0)(1)));
+    for (int i = 0; i < 16; i += 4) {
+        h->mov(4, tmp.ud(0)(1), max.ud(i)(1));
+        h->sel(4 | ge, max.f(0), max.f(0)(1), tmp.f(0)(1));
+    }
+    h->mov(2, tmp.ud(0)(1), max.ud(2)(1));
+    h->sel(2 | ge, max, max.f(0)(1), tmp.f(0)(1));
+    h->mov(2, tmp.ud(0)(1), max.ud(1)(1));
+    h->sel(1 | ge, max, max.f(0)(1), tmp.f(0)(1));
```

Comment:
Would this work:
```suggestion
    h->sel(8 | ge, max.ud(0)(1), max.ud(0)(2), max.ud(1)(2));
    h->sel(4 | ge, max.ud(0)(1), max.ud(0)(2), max.ud(1)(2));
    h->sel(3 | ge, max.ud(0)(1), max.ud(0)(2), max.ud(1)(2));
    h->sel(1 | ge, max.ud(0)(1), max.ud(0)(0), max.ud(1)(0));
```
(using the fact that all of the data are positive floats, so they sort the same as unsigned ints)

Reply:
I guess there'd need to be some special handling of nan/infs before this sequence to avoid propagating them, but similarly, the existing sequence needs special inf handling so _they_ don't propagate and we end up with a bad scale.

Reply:
@atkassen -- you can add 0x80400000:ud to the inputs prior to the max, and then subtract after the max.

Another optimization to consider (maybe not in this PR) is a fully vectorized horizontal reduction, where you recombine partly reduced vectors as you go, so that you can get full SIMD usage at each stage -- vISA example [here (DEFINE_HREDUCE16_FLOAT)](https://github.com/intel/sycl-tla/pull/547/commits/9a6aa27c3307e82d41839c2747c3043827cee749).